xarray/core/dataset.py

```diff
@@ -5723,16 +5723,17 @@ def _assert_all_in_dataset(
 
     def drop_vars(
         self,
-        names: Hashable | Iterable[Hashable],
+        names: Hashable | Iterable[Hashable] | Callable,
```

Comment:
```suggestion
        names: Hashable | Iterable[Hashable] | Callable[[Self], Hashable | Iterable[Hashable]],
```
You could try this :)

And not to mention the old `str | Iterable[Hashable]` story. But the code below seems fine with the current definition, so we can leave it.